### ISSUE TITLE
Integrate Collaborative Editing to LAPCE

### ISSUE THREAD

**Issue by tafaB** [lapce/lapce#2259]
Let's delve into the exciting world of collaborative editing and its potential integration into LAPCE!

**Comment by hbina** [comment]
I would like to know more about what you mean by collaborative editing? Do you mean something like this https://swimm.io/ or like Google Docs with multiple concurrent writers.

**Comment by panekj** [comment]
> I would like to know more about what you mean by collaborative editing? Do you mean something like this [swimm.io](https://swimm.io/) or like Google Docs with multiple concurrent writers.

Pretty sure it's Live Share or what Zed provides

**Comment by tafaB** [comment]
Yes, something like Live Share and Google Docs where multiple users edit the same file
I have already studied the algorithms and data types need, but the integration in LAPCE is what is challenging right now (mostly because I am new to RUST)
Currently trying to implement the modification of the just one file (later on a workspace may be shared)
First I thought on adding it as a plugin, but got informed the there is not much that can be done by plugins right now in LAPCE (mostly lsp and colorschemes). Now I am trying to add it to the list of Panels.

**Comment by panekj** [comment]
I meant the code responsible for interaction with editor content

**Comment by tafaB** [comment]
> I meant the code responsible for interaction with editor content

Understood. What about the interface of creating a collaboration session or joining one?

**Comment by panekj** [comment]
it would be in data and ui

**Comment by tafaB** [comment]
> it would be in data and ui

Understood. Thank you a lot.

**Comment by tafaB** [comment]
Hello, I implemented a new panel for "collaboration" (based on the other panels, ex: source control) and all the needed dependencies, but still the panel is not showing up.
Could you please let me know what is the part of the source code ex: for source control, that is responsible for showing the panel (icon and content).

**Comment by hbina** [comment]
Are you talking about one of these here? 
![Screenshot from 2023-03-21 19-39-05](https://user-images.githubusercontent.com/6733660/226595490-a2ed4d4d-7152-4f37-a90e-a2e5799635cb.png)

**Comment by poelzi** [comment]
I would love such a feature.
Seee https://github.com/ekzhang/rustpad
And https://docs.rs/operational-transform/latest/operational_transform/
There is a lib that implements proper merging for distributed editing.
It would also be nice if a second user could join a master instance that provides the build toolchain and the client can open all the project files and edit files but does not run the toolchain. The master can also set clients readonly would be cool.

**Comment by tafaB** [comment]
> I would love such a feature.
> Seee https://github.com/ekzhang/rustpad
> And https://docs.rs/operational-transform/latest/operational_transform/
> There is a lib that implements proper merging for distributed editing.
> It would also be nice if a second user could join a master instance that provides the build toolchain and the client can open all the project files and edit files but does not run the toolchain. The master can also set clients readonly would be cool.

Yes, I have studied the theory behind it (OT and CRDTs) and firstly I am trying to achieve the collaboration in just one file .. later on a complete project may be achieved.

**Comment by hbina** [comment]
> > Are you talking about one of these here?
> > ![Screenshot from 2023-03-21 19-39-05](https://user-images.githubusercontent.com/6733660/226595490-a2ed4d4d-7152-4f37-a90e-a2e5799635cb.png)
> 
> yes the list: FileExplorer SourceControl Plugin Terminal Search Problem

Its a bit complicated because panels are dynamic. But you can take a look at the `LapcePanel` here. https://github.com/lapce/lapce/blob/53cd9a8051d0d2e2f5a01723e481e435e5063021/lapce-ui/src/panel.rs#L85

**Comment by The-Minecraft-Scientist** [comment]
to do this "properly," we'd need to make changes to the CRDT implementation in the xi-editor Rope data type that actually stores the text in a buffer, as mentioned [here](https://xi-editor.io/docs/rope_science_08a.html). The Rope code currently lives [in this xi-editor fork](https://github.com/lapce/xi-editor/tree/master/rust/rope).

**Comment by niklaskorz** [comment]
It would be great if this were to follow the https://github.com/teamtype/teamtype protocol for interop with other editor plugins.
